Add Sungrow solar inverter integration

DasBasti opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Add modbus read out of Sungrow solar inverter. 
Tested with SG5.0RT but should work with:
SG30KTL-M, SG30KTL-M-V31, SG33KTL-M, SG36KTL-M, SG33K3J, SG49K5J, SG34KJ, LP_P34KSG, SG50KTL-M-20, SG60KTL, SG80KTL, SG80KTL-20, SG60KU-M, SG5KTL-MT, SG6KTL-MT, SG8KTL-M, SG10KTL-M, SG10KTL-MT, SG12KTL-M, SG15KTL-M, SG17KTL-M, SG20KTL-M, SG80KTL-M, SG111HV, SG125HV, SG125HV-20, SG33CX, SG40CX, SG50CX, SG110CX, SG30KTL, SG10KTL, SG12KTL, SG15KTL, SG20KTL, SG30KU, SG36KTL, SG36KU, SG40KTL, SG40KTL-M, SG50KTL-M, SG60KTL-M, SG60KU

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/21472

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

DasBasti committed:
Add intergration for Sungrow solar inverter

homeassistant commented:
Hi @DasBasti, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#71526).

Once you do that we will be able to review and accept this pull request.

Thanks!

DasBasti committed:
Add testcase for 'unknown error'

DasBasti committed:
Add test cases for the sungrow sensor

DasBasti committed:
Add testcase for sungrow data update connection failure

DasBasti committed:
Refine check of input values

DasBasti committed:
Additional tests to get coverage to 100%

DasBasti commented:
I think I managed to fulfill all the requirements in the tests. Could you please trigger a run?

davet2001 commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+    if "4990 ~ 4999 - Serial number" not in client.inverter:
+        raise InvalidAuth
+
+    # If you cannot connect:
+    # throw CannotConnect
+    # If the authentication is wrong:
+    # InvalidAuth
```
Please remove commented out code.

davet2001 commented on homeassistant/components/sungrow/const.py:
```diff
@@ -0,0 +1,181 @@
+        icon="mdi:solar-power",
+        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        value=lambda value: round(value, 3),
```
You should probably set the state class to INCREASING here (this enables it to appear in the energy metering features in HA)

DasBasti replied:
Is TOTAL_INCREASING something for the daily run time? That is the amount of hours the system was online each day.
Summary values are provided by the "running time total" and "yield total".
These two summaries would be different if the device is not connected from the beginning if I understand it correctly.

davet2001 replied:
I am not 100% sure.

davet2001 reviewed (changes requested):
Initial review to get started.

DasBasti committed:
Remove commented out code.

davet2001 commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+            errors["base"] = "cannot_connect"
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
```
Probably this is not necessary, as there are only a few errors you need to catch.

davet2001 commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
```
You can put this at the end, then you don't need to duplicate it.

I.e.
```
errors = {}

if user_input is not None:
	[handle user input and eventually create_entry if data is good]

[show form]
````

davet2001 commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+        )
+
+
+class CannotConnect(HomeAssistantError):
```
I couldn't see what used this.

davet2001 commented on homeassistant/components/sungrow/const.py:
```diff
@@ -0,0 +1,181 @@
+    ),
+    # '5017 - Total DC power' - '5031 - Total active power'
+    SungrowSensorEntityDescription(
+        key="alternator_loss",
```
Is this correct? Doesn't seem to fit with the comment or the format of the other keys

DasBasti replied:
alternator loss is calculated from the two values mentioned in the comment. This is done when the sensor values are updated.

davet2001 replied:
ok, yes i see now. :)

davet2001 commented on homeassistant/components/sungrow/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "name": "Sungrow",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/sungrow",
+  "requirements": ["sungrow-modbus-client==0.0.2", "pymodbus==2.5.3"],
```
I saw several other sungrow modus TCP clients, that seemed to have been around for a while already.

Is there a reason you chose to roll your own in this case?

davet2001 replied:
https://pypi.org/project/SungrowModbusTcpClient/
https://github.com/meltaxa/solariot

DasBasti replied:
I also found those two projects. But they didn't fit the needs I was having.
The SungrowModbusTcpClient does not provide a way to interpret the actual values and solariot is not a library but rather a service to load the information based on a config file.

davet2001 replied:
Ok 😃

davet2001 commented on tests/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,74 @@
+    "5024 - Phase C current": 4.1,
+    "5031 - Total active power": 2940,
+    "5033 - Total reactive power": 6,
+    "5035 - Power factor": 1000,
```
I thought power factors were always between 0 and 1?  If so maybe the test should be a decimal number?

DasBasti replied:
The power factor value is returned from the device with a factor 1000. The test data is what would be returned from the client library.

davet2001 commented on tests/components/sungrow/test_config_flow.py:
```diff
@@ -0,0 +1,167 @@
+    assert result["errors"] is None
+
+    with patch(
+        "homeassistant.components.sungrow.config_flow.validate_input",
```
If you are testing your form, is it possible to patch the return value of the library rather patch the form validation functions (might be needed for test coverage anyway)?

DasBasti replied:
I created a MockClient for testing the form.

davet2001 reviewed (changes requested):
A few more comments, I need to review the last part of the tests a bit more though.

DasBasti commented on homeassistant/components/sungrow/const.py:
```diff
@@ -0,0 +1,181 @@
+    # '5035 - Power factor'
+    SungrowSensorEntityDescription(
+        key="5035 - Power factor",
+        name="efficiency",
```
i tried to align the names with the solarlog component. power factor would be fine with me.

DasBasti commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,91 @@
+        )
+
+        self.host = entry.data[CONF_HOST]
+        self.port = 502
+        if CONF_PORT in entry.data:
+            self.port = entry.data[CONF_PORT]
```
will do that

DasBasti commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,91 @@
+            data["5017 - Total DC power"] - data["5031 - Total active power"]
+        )
+
+        data["time"] = homeassistant.util.dt.utcnow()
```
ok, that this is unnecessary

DasBasti commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+    client = Client(sungrow_register.regmap, data["host"], port)
+
+    client.load_register()
+    if "4990 ~ 4999 - Serial number" not in client.inverter:
+        raise InvalidAuth
```
if no auth is needed I can freely access all values. I picked one at random to determine if access was successful.

davet2001 replied:
Yes, better to pick an exception that fits the actual problem.  It's important because some of the exceptions turn into error messages that users will see on the UI

DasBasti committed:
Implement recommended changes

Simplyfy code in SungroData.
Add correct return types to update functions.
Remove time element.
Testing with mock library.

DasBasti commented on homeassistant/components/sungrow/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+    port = data.get("port", 502)
+    client = Client(sungrow_register.regmap, data["host"], port)
+
+    client.load_register()
```
Ah ok. Basically the same as I did fetching the sensor data periodically

davet2001 replied:
Not sure where you mean but yes, that is the principle, assuming you are using `await`.

DasBasti replied:
Ok, I changed it to use the async executer job.

DasBasti committed:
Call blocking function asyncronically

davet2001 commented on tests/components/sungrow/test_init.py:
```diff
@@ -0,0 +1,79 @@
+        entry = create_entry(hass)
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+        await hass.data[DOMAIN][entry.entry_id].async_refresh()
```
I was expecting some sort of validation that new data is actually present after the refresh.

DasBasti replied:
I added more validation tests for this testcase.

davet2001 replied:
Thank you, looks good

davet2001 reviewed (changes requested):
Looking good.  Thanks for covering previous points. Just spotted one thing when reviewing the last part of the tests.

DasBasti committed:
Add more assertions for testing data update

davet2001 commented on tests/components/sungrow/test_config_flow.py:
```diff
@@ -0,0 +1,103 @@
+
+
+async def test_validate_input(hass: HomeAssistant) -> None:
+    """Test Input validation with valid auth."""
```
Where is the authentication?  Is it part of the 'host' field?

davet2001 commented on tests/components/sungrow/test_init.py:
```diff
@@ -0,0 +1,99 @@
+        await hass.data[DOMAIN][entry.entry_id].async_refresh()
+
+        assert (
+            hass.data[DOMAIN][entry.entry_id].data["4990 ~ 4999 - Serial number"]
```
Consider defining these special strings as constants.

DasBasti replied:
I did that

davet2001 reviewed (changes requested):
Thanks for the edits so far, sorry it's taken a long time for 2nd review.  Some minor issues identified but looking good!

DasBasti committed:
Sungrow: move key strings to constants

DasBasti committed:
Sungrow: Use constants instead of strings for keys

DasBasti committed:
Sungrow: pymodbus is not a requirement

DasBasti committed:
Sungrow: allow only unique serial numbers

DasBasti committed:
Sungrow: remove unused strings

DasBasti committed:
Sungrow: clarify comment in test case

DasBasti committed:
Sungrow: Test update deprecated result types

DasBasti committed:
Sungrow: Use state getter to test sensor data

DasBasti commented on tests/components/sungrow/test_init.py:
```diff
@@ -0,0 +1,99 @@
+        assert (
+            hass.data[DOMAIN][entry.entry_id].data["4990 ~ 4999 - Serial number"]
+            == inverter_data["4990 ~ 4999 - Serial number"]
+        )
```
I used the states object to get sensor data to compare against the data given by the mocked device.

davet2001 replied:
Sure, this works, although as I said it is less common, and doesn't verify that the sensor got created with the name/domain you expected.

DasBasti commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,87 @@
+        url = urlparse(self.host, "http")
+        netloc = url.netloc or url.path
+        path = url.path if url.netloc else ""
+        url = ParseResult("http", netloc, path, *url[3:])
```
I reduced the string manipulation to a simpler version. This was unnecessarily convoluted.

DasBasti replied:
the reduced string manipulation was not able to take into account having "http://1.1.1.1" or "1.1.1.1" as input. so i added comments

davet2001 replied:
I think that the comment should say something like "Allow user to enter URLs with or without 'http://' at the start", if that is the intention of the code.

davet2001 replied:
I also recommend a parametrised pytest for this so that you can verify that the different URL schemes you want to accept all still work.

But that might be best in a follow up PR.

DasBasti commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,87 @@
+
+        self.client = Client(sungrow_register.regmap, self.host, self.port)
+
+        self.logger.debug(
```
did that

DasBasti commented on homeassistant/components/sungrow/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
+  "after_dependencies": ["http"],
```
I removed unused dependencies from the manifest

DasBasti commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,87 @@
+            _LOGGER.error("No route to host/endpoint: %s", self.host)
+            raise update_coordinator.UpdateFailed(err)
+
+        data["alternator_loss"] = data.get("5017 - Total DC power", 0) - data.get(
+            "5031 - Total active power", 0
+        )
```
all magic strings are converted to constants.

DasBasti commented on homeassistant/components/sungrow/strings.json:
```diff
@@ -0,0 +1,20 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
```
I removed the unused error strings

DasBasti commented on requirements_all.txt:
```diff
@@ -1655,6 +1655,7 @@ pymitv==1.4.3
 
 # homeassistant.components.modbus
+# homeassistant.components.sungrow
 pymodbus==2.5.3
```
I removed the unused dependencies

DasBasti committed:
Sungrow: replace more key strings

DasBasti committed:
Sungrow: use correct logger

DasBasti committed:
Sungrow: simplify url string parsing

DasBasti committed:
Sungrow: directly access host field

davet2001 commented on homeassistant/components/sungrow/__init__.py:
```diff
@@ -0,0 +1,87 @@
+        self.host = entry.data.get(CONF_HOST, "")
+        self.port = entry.data.get(CONF_PORT, 502)
+
+        self.unique_id = entry.entry_id
```
I see you are now using a unique id coming from the device serial number.

davet2001 commented:
I added a link to your documentation PR in the description.

DasBasti committed:
Revert "Sungrow: simplify url string parsing"

This reverts commit 5aaa0432af6b312ba9c82830dc5541a5b2e8debc.

DasBasti committed:
Sungrow: add comments to string mainpulation

The mainpulation is necessary to cover addresses with and without
http prefix.

DasBasti committed:
Sungrow: clarify comment

DasBasti committed:
Sungrow: Tests now test different address types

The test now covers addresses with http/https and without prefix.

davet2001 commented on tests/components/sungrow/test_init.py:
```diff
@@ -7,27 +7,27 @@
-    assert entry.state is ConfigEntryState.NOT_LOADED
-    assert not hass.data.get(DOMAIN)
+        assert entry.state is ConfigEntryState.NOT_LOADED
+        assert not hass.data.get(DOMAIN)
```
````suggestion
@pytest.mark.parametrize(
    "host",
    [
        "1.1.1.1",
        "http://1.1.1.1",
        "https://1.1.1.1",        
    ],
)
async def test_setup(hass: HomeAssistant, host) -> None:
    """Test creation and unload with different address variants."""
    entry = create_entry(hass, host, 502)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()
    assert entry.state == ConfigEntryState.LOADED

    assert await hass.config_entries.async_unload(entry.entry_id)
    await hass.async_block_till_done()

    assert entry.state is ConfigEntryState.NOT_LOADED
    assert not hass.data.get(DOMAIN)
````

davet2001 replied:
`@pytest.mark.parametrize` does this type of test for you.

DasBasti replied:
Thank you. Didn't know that.

DasBasti committed:
Sungrow: use pytest parameterize for addresses

Co-authored-by: Dave T <17680170+davet2001@users.noreply.github.com>

DasBasti committed:
Sungrow: Fix test with address variants

The previous commit removed the imports at the top of the test.

DasBasti commented:
The code change merged by github somehow mangled the import section of the test. The last commit fixes that.

DasBasti commented:
Is there still something missing for the merge?

davet2001 commented:
The pytests are currently failing.  Please could you check why?

DasBasti commented:
The tests failed in an unrelated component (history_stats/test_sensor.py::test_measure_sliding_window) and might have been realted to #81175. I rebased the PR to update to the current code.

DasBasti commented:
This time the jobs fail in cache cleanup and recorder component tests. What is the correct way to deal with such issues?

davet2001 commented:
Sorry that there are problems causing difficulty here.

If there are flaky/failing CI tests on the dev branch then probably you are not the only one to notice.  

Rebase and retest in 24h often works in that case.

But I had a check on discord and couldn't see it mentioned.

davet2001 commented on tests/components/sungrow/test_config_flow.py:
```diff
@@ -0,0 +1,103 @@
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_validate_input(hass: HomeAssistant) -> None:
```
Probably this test is already done by your other tests.

davet2001 reviewed (approved):
Looks good to me! Thanks @DasBasti :)

02JanDal committed:
Change library used for Sungrow

02JanDal commented:
@davet2001 I've been in contact with @DasBasti as we've both been working on an integration for Sungrow inverters in parallel. As he was further ahead with the integration itself and the process of getting it reviewed and merged while my backing library was more advanced we decided to work together already as part of this PR, since it was not yet merged. As such I have modified Bastians code to use my library.

A review of these changes would be greatly appreciated!

davet2001 commented:
It looks like you have rebased and force pushed.

I can't easily see what has changed since I last reviewed and I would prefer not to re-read all >1000 lines.  

Please could you describe what you changed?

02JanDal commented:
@davet2001 Thanks for taking the time!

Here is a diff of just the changes to this integration: https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282
Note that it excludes changes in generated files like requirements and CODEOWNERS, as those also include changes from other commits so are harder for me to isolate.

The file with the most significant changes is [`__init__.py`](https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282#file-sungrow-diff-L1-L284), as my library is async (although not 100% until HA switches to pymodbus 3 but that has to wait for upstream and the modbus integration) I had to move some stuff out of the SungrowData constructor into `async_setup_entry`. It also contains some aspects in `SungrowCoordinatorEntity` previously found in `sensor.py`, which is mostly just a preparation for more entity types (which I've got ready locally but will be part of a future PR).

[`config_flow.py`](https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282#file-sungrow-diff-L285-L368) should be functionally equivalent.

[`const.py`](https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282#file-sungrow-diff-L369-L588) lost the sensor definitions, I moved them into `sensor.py`, again as it makes more sense to me once more entity types are there.

[`sensor.py`](https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282#file-sungrow-diff-L608-L1075) gained the sensor definitions, quite a few more as my library supports more, while also losing some details in them (like device class and unit) that I instead compute inside of `SungrowCoordinatorEntity` based on the metadata in `pysungrow`.

In the tests the changes are almost exclusively about changing what is mocked, I did however introduce a [fixture](https://gist.github.com/02JanDal/7c0d50e191e6bc74422a6c81ac6e1282#file-sungrow-diff-L1207-L1231) that takes care of the patching that is done for most tests, as well as creating and registering the config entry.

Other than that it's of course quite a few import changes, as well as some adaptions to changes in the HA API, like using `SensorDeviceClass` instead of strings and going from `async_setup_platforms` to `async_forward_entry_setups`.

davet2001 commented:
Ok, this is quite a big job now.

It's close to a rewrite and using quite a new python library (so maybe that needs reviewing as well).

There is a big benefit to keeping a PR small. Faster to review, faster to re-review etc. Will re-review as soon as possible but it won't be immediately.

edenhaus commented on homeassistant/components/sungrow/translations/de.json:
```diff
@@ -0,0 +1,20 @@
+{
```
Please remove the translations folder. All translation strings should go into `strings.json`

edenhaus commented on homeassistant/components/sungrow/sensor.py:
```diff
@@ -0,0 +1,436 @@
+            or key.startswith("daily_")
+            or key.startswith("monthly_")
+        ) and variable.unit == "kWh":
+            self._attr_state_class = SensorStateClass.TOTAL_INCREASING
```
As you already use the SensorEnityDescription, please add the state class directly there.

edenhaus commented on homeassistant/components/sungrow/sensor.py:
```diff
@@ -0,0 +1,436 @@
+    )
+
+
+class SungrowSensor(SungrowCoordinatorEntity, SensorEntity):
```
New integrations/entities must support [translation of names and attributes](https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/).

Please adopt your code accordingly.

edenhaus reviewed (changes requested):
Please rebase and fix the CI

DasBasti committed:
Sungrow: remove transaltion folder

Collect all strings in strings.json

DasBasti committed:
Sungrow: Manifest keys have been sorted: domain, name, then alphabetical order

DasBasti committed:
Sungrow: Use translation of names and attributes

DasBasti commented:
@02JanDal can you please have a look into the changes I made, since I currently have no access to a sungrow inverter anymore

02JanDal commented:
Yes, I need to make another push on this, need to finish the update to pymodbus 3 (which the modbus integration was updated to) and fix a few instabilities regarding the various ways to connect (WiFi, LAN, LAN via WiNet stick) and bugs found by a few testers in a Swedish group where I posted this. Hope to get to this early next week.

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
